// tapper-core/dist/commands/tapper.command.query.builder.js
export class TapperCommandQueryBuilder {
    queryString = "adb";
    setSystemSettingsKey(key) {
        this.queryString += " " + key;
        return this;
    }
    setGeneralSettingsKey(key) {
        this.queryString += " " + key;
        return this;
    }
    setAndroidTestingKey(key) {
        this.queryString += " " + key;
        return this;
    }
    setShellCommand() {
        this.queryString += " shell";
        return this;
    }
    setSystem() {
        this.queryString += " system";
        return this;
    }
    setProp() {
        this.queryString += " setprop";
        return this;
    }
    setSettings() {
        this.queryString += " settings";
        return this;
    }
    put() {
        this.queryString += " put";
        return this;
    }
    setGlobal() {
        this.queryString += " global";
        return this;
    }
    setBoolean(status) {
        if (status) {
            this.queryString += " true";
        }
        else {
            this.queryString += " false";
        }
        return this;
    }
    setEnabledByNumbers(status) {
        if (status) {
            this.queryString += " 1";
        }
        else {
            this.queryString += " 0";
        }
        return this;
    }
    setInput() {
        this.queryString += " input";
        return this;
    }
    setEnabledByNumbersNextLevel(status) {
        if (status) {
            this.queryString += " 2";
        }
        else {
            this.queryString += " 1";
        }
        return this;
    }
    setCustomValue(value) {
        this.queryString += " " + value;
        return this;
    }
    getQuery() {
        return this.queryString;
    }
}
