// tapper-core/dist/models/android.general.settings.key.js
export var AndroidGeneralSettingsKey;
(function (AndroidGeneralSettingsKey) {
    AndroidGeneralSettingsKey["ToggleDarkMode"] = "night_display_activated";
    AndroidGeneralSettingsKey["ToggleWifi"] = "wifi_network_suggestions_splice";
    AndroidGeneralSettingsKey["Devices"] = "devices";
    AndroidGeneralSettingsKey["DetailedDevices"] = "devices -l";
    AndroidGeneralSettingsKey["InstallApk"] = "install";
    AndroidGeneralSettingsKey["UnInstallApk"] = "uninstall";
    AndroidGeneralSettingsKey["Reboot"] = "reboot";
    AndroidGeneralSettingsKey["RemovePermissions"] = "reset-permissions";
    AndroidGeneralSettingsKey["PowerSavingMode"] = "low_power";
    AndroidGeneralSettingsKey["HomeButton"] = " shell input keyevent 3";
    AndroidGeneralSettingsKey["PhoneNumber"] = " shell am start -a android.intent.action.CALL -d tel:";
    AndroidGeneralSettingsKey["OpenUrl"] = " shell am start -a android.intent.action.VIEW -d";
    AndroidGeneralSettingsKey["Screenshot"] = "shell screencap -p";
    AndroidGeneralSettingsKey["OpenApp"] = "shell monkey -p {Package} -c android.intent.category.LAUNCHER 1";
})(AndroidGeneralSettingsKey || (AndroidGeneralSettingsKey = {}));
