// tapper-core/dist/utils/tapper.command.execution.manager.js
import { exec } from 'child_process';
/**
 * Any Shell Command Should be Executed Here In This File
 * Helpful in Debug, if you want to See Which Query Submitted and How Add a breakpoint Here
 */
export class TapperCommandExecutionManager {
    // @ts-ignore
    static onExecuteCommandString(command) {
        exec(command, (error, stdout, stderr) => {
            if (error) {
                console.log("Error While Executing the Command ... " + error.message);
                return;
            }
            console.log("Command To Execute: " + command);
            console.log(stdout);
        });
    }
    // @ts-ignore
    static onExecuteCommandByDeviceType(command) {
        exec(command.command, (error, stdout, stderr) => {
            if (error) {
                console.log("Error While Executing the Command ... " + error.message);
                return;
            }
            const output = stdout.trim().replace("\n", "").replace("\t", "");
            if (output) {
                console.log(command.name + "=" + output);
            }
        });
    }
    static getRandomCoordinates(screenWidth, screenHeight) {
        const minX = 0;
        const maxX = screenWidth;
        const minY = 0;
        const maxY = screenHeight;
        const randomX = Math.floor(Math.random() * (maxX - minX + 1)) + minX;
        const randomY = Math.floor(Math.random() * (maxY - minY + 1)) + minY;
        return { x: randomX, y: randomY };
    }
    static sleep(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }
    static onExecuteCommandStringWithoutInput(command) {
        exec(command, (error, stdout, stderr) => {
            if (error) {
                console.log("Error While Executing the Command ... " + error.message);
                return;
            }
        });
    }
}
